Guide for AA-only workflows (#270)
nextstrain/docs.nextstrain.org: src/guides/bioinformatics/amino-acid-workflows.rst

====================
Amino Acid Workflows
====================

Almost all Augur workflows describe nucleotide analyses. Here we describe
workflows using entirely Amino Acid (AA) sequences. This guide assumes
familiarity with general (nucleotide!) Augur workflows.

This guide first explains the changes needed to analyse a set of (unaligned)
AA sequences representing a single gene/CDS, including ancestral reconstruction.
We then describe how you can :ref:`concatenate multiple genes together for increased resolution <multiple-gene-analysis>`.
Finally there's a :ref:`summary of the differences in Auspice behaviour when viewing protein-only datasets <auspice>`.

Prerequisites
=============

* Metadata TSV
* AA sequence FASTA (unaligned or aligned)
* Reference sequence for alignment, if not included in your sequences. FASTA or GenBank format. 
* Augur version TKTK (TODO XXX needs release)

> * Auspice version TKTK (TODO XXX needs release)

Author: (reminder for me)

This guide first explains the changes needed to analyse a set of (unaligned)
AA sequences representing a single gene/CDS, including ancestral reconstruction.
We then describe how you can :ref:`concatenate multiple genes together for increased resolution <multiple-gene-analysis>`.
Finally there's a :ref:`summary of the differences in Auspice behaviour when viewing protein-only datasets <auspice>`.

Prerequisites
=============

* Metadata TSV
* AA sequence FASTA (unaligned or aligned)
* Reference sequence for alignment, if not included in your sequences. FASTA or GenBank format. 
* Augur version TKTK (TODO XXX needs release)
* Auspice version TKTK (TODO XXX needs release)

Changes compared to a nucleotide workflow
=========================================

Alignment
---------

``augur align --method mafft`` (the default) will work for protein alignments.
Note that Nextclade doesn't support AA inputs and thus can't be used as an aligner in this case.

Tree construction
-----------------

IQ-TREE can be parameterized for AA sequences by adding the the following arguments to
``augur tree``:

* ``--method iqtree`` (the default)
* ``--tree-builder-args '--seqtype AA'``

> * ``--substitution-model LG`` (the default IQ-TREE substitution model). See the "Specifying substitution models" section of the `IQ-TREE docs <https://iqtree.github.io/doc/Command-Reference>`__ for a full list of available models.

Reviewer: In general, I find these guides easier to follow when the full command is spelled out in a code-block that I can copy/paste to run instead of listing out the different options. Or, since this section specifically says changes compared to NT workflows, it'd be nice to see the side-by-side comparison of the NT command vs the AA command.

Author: I've added snippets to each section for a basic flu-based workflow, but in general I am wary of users copy-pasting stuff and want to encourage them to understand the components involved :)

Temporal inference (refine)
---------------------------

Augur refine will handle AA alignments if provided with the ``--aa`` flag.

> Internally this changes TreeTime to use the Jones-Taylor-Thornton (JTT92)

Reviewer: For those unfamiliar, could `Jones-Taylor-Thornton (JTT92)` link to a relevant paper? (I think https://doi.org/10.1093/bioinformatics/8.3.275?)

Author: I've linked to TreeTime. I don't think that 1992 paper is going to help all that much for someone who doesn't know beforehand what JTT is!

substitution model. You may also want to use ``--branch-length-inference
marginal`` if your tree has long branches.

Ancestral reconstruction
------------------------

For nucleotide analyses there are two main approaches to reconstructing
amino-acid sequences and thus inferring AA mutations across the tree: ``augur
ancestral`` which independently translates provided nuc & AA alignments, and
``augur translate`` which translates nucleotide sequences reconstructed across
the tree.

``augur ancestral`` can work for AA-only analyses. You will need to provide the following options to reconstruct a single gene/CDS:

* ``--genes <GENE NAME>``
* ``--translations <FILENAME>`` (amino acid FASTA) the filename can be a specific path or include a ``%GENE`` placeholder which will be replaced with the gene name
* [optional] ``--root-sequence <FILENAME>`` (amino acid FASTA) the filename can be a specific path or include a ``%GENE`` placeholder which will be replaced with the gene name

> * [optional] ``--annotation <GenBank or GFF>`` (not needed for a single gene analysis)

Reviewer: It'd be nice to have these options spelled out in a full command with a tree of the example files to fully explain what the `%GENE` replacement means.

See :doc:`./root-and-ref-seqs` for more information about how providing a
root-sequence will assign mutations to the root-node of the tree.

The optional annotation file allows the actual nucleotide coordinates of the
gene to be used in the resulting JSON, which slightly changes how the
dataset will appear in Auspice (see :ref:`below <auspice>`).

Other commands (export, traits etc)
-----------------------------------

> Other Augur commands in a typical workflow should work with protein sequences.

Reviewer: This got me thinking if `augur filter` would work with AA sequences. The metadata filters should work as expected, but maybe we need to revisit the [sequences filters](https://docs.nextstrain.org/projects/augur/en/stable/usage/cli/filter.html#augur-make_parser-sequence-filters)?

Author: Ahhh, I forgot about sequence filtering. I'll take a look. Thanks!

Author: This became a rabbit hole, but is now done. Changes to `augur index`, `augur filter`, `augur subsample`

.. _multiple-gene-analysis:

Concatenating multiple genes together for increased resolution
==============================================================

For increased resolution you may wish to concatenate multiple genes/CDSs
together, but still infer the ancestral sequences independently for each gene.
The main difference from the workflow described above is the need for sequence
concatenation. Augur doesn't provide a tool for this directly but the following
Snakemake rule may be a good starting point:

.. code-block:: text

    rule combine_alignments:
        input:
            alignments = ["<PATH>", "<PATH>", ...]
        output:
            alignment = "<PATH>"
        run:
            from Bio import SeqIO
            alignments = [SeqIO.to_dict(SeqIO.parse(f, "fasta")) for f in input.alignments]
            names = [set(a.keys()) for a in alignments]
            if not all(n == names[0] for n in names[1:]):
                raise ValueError("Sequence names differ between alignments.")
            with open(output.alignment, "w") as fh:
                for name in alignments[0]:
                    seq = "".join(str(aln[name].seq) for aln in alignments)

>                     fh.write(f">{name}\n{seq}\n")

Reviewer: I think this can be done with [seqkit concat](https://bioinf.shenwei.me/seqkit/usage/#concat)

```suggestion
shell:
    """
    seqkit concat {input.alignments} > {output.alignment}
    """
```

Author: Thanks!

        --annotation HA_genemap.gff \
        --genes HA1 HA2 \
        --translations alignment_%GENE.fasta \
        --root-sequence root-sequence_%GENE.fasta \
        --output-node-data muts.json

.. _auspice:
        
Visualisation in Auspice
========================

For the most part protein-only analyses work in Auspice as expected. The tree divergence (branch lengths)
will be expressed either in units of amino acid mutations or amino acid substitutions per codon site, and
of course there won't be any nucleotide mutations annotated on branches!

For analyses which reconstruct mutations (`augur ancestral`) for a single gene, the entropy panel will
only show that gene (Figure 1). The nucleotide coordinates mapping to this gene differ slightly depending
on whether an annotation (genemap) was provided to `augur ancestral`.

.. image :: ../../images/entropy-panel-single-gene.png
   :alt: Entropy panel for a single gene analysis

*Figure 1: Two screenshots of how a single gene analysis may be displayed in
Auspice. The teal screenshot is from an analysis using a genemae so that the
nucleotide axis (the lower of the two teal axes) uses the true nucleotide

> coordinates of HA1: 1053..1715nt. The blue screenshot is from an analysis

Reviewer: ```suggestion
coordinates of HA1: 66..1052nt. The blue screenshot is from an analysis
```

Author: Oh what a mess - I changed to using HA2 but everything's a little out of sync. Let me tidy up.

====================
Amino Acid Workflows
====================

Almost all Augur workflows describe nucleotide analyses. Here we describe
workflows using entirely Amino Acid (AA) sequences. This guide assumes
familiarity with general (nucleotide!) Augur workflows.

This guide first explains the changes needed to analyse a set of (unaligned)
AA sequences representing a single gene/CDS, including ancestral reconstruction.
We then describe how you can :ref:`concatenate multiple genes together for increased resolution <multiple-gene-analysis>`.
Finally there's a :ref:`summary of the differences in Auspice behaviour when viewing protein-only datasets <auspice>`.

.. note::

   For deeply diverged proteins you should use other approaches, and pay particular
   attention to model selection. Our temporal inference & tree refinement approach is
   not well suited to very long branches.

Prerequisites
=============

* Metadata TSV
* AA sequence FASTA (unaligned or aligned)
* Reference sequence for alignment, if not included in your sequences. FASTA or GenBank format. 
* Augur version TKTK (TODO XXX needs release)

> * [Auspice version 2.69.0](https://github.com/nextstrain/auspice/releases/tag/v2.69.0)

Reviewer: ```suggestion
* `Auspice version 2.69.0 <https://github.com/nextstrain/auspice/releases/tag/v2.69.0>`__
```

Changes compared to a nucleotide workflow
=========================================

Alignment
---------

``augur align --method mafft`` (the default) will work for protein alignments.
Note that Nextclade doesn't support AA inputs and thus can't be used as an aligner in this case.

Tree construction
-----------------

IQ-TREE can be parameterized for AA sequences by adding the the following arguments to
``augur tree``:

* ``--method iqtree`` (the default)
* ``--tree-builder-args '--seqtype AA'``
* ``--substitution-model <MODEL>`` The default IQ-TREE substitution model is ``LG``; see the "Specifying substitution models" section of the `IQ-TREE docs <https://iqtree.github.io/doc/Command-Reference>`__ for a full list of available models.

An example command to construct a tree from a concatenated HA1 + HA2 alignment may look like:

.. code-block:: shell

   augur tree
     --method iqtree --substitution-model LG --tree-builder-args '--seqtype AA' \

>      --alignment results/HA12/aligned.fasta  --output results/HA12/tree_raw.nwk

Reviewer: Small preference for one option per line to make the code more readable

```suggestion
   augur tree \
     --method iqtree \
     --substitution-model LG \
     --tree-builder-args '--seqtype AA' \
     --alignment results/HA12/aligned.fasta  \
     --output results/HA12/tree_raw.nwk
```

Temporal inference (refine)
---------------------------

Augur refine will handle AA alignments if provided with the ``--aa`` flag.
Internally this changes `TreeTime to use the Jones-Taylor-Thornton (JTT92)
substitution model
<https://treetime.readthedocs.io/en/latest/tutorials/ancestral.html#amino-acid-sequences>`__.
You may also want to use ``--branch-length-inference marginal`` depending on the
branch lengths involved.

An example command may look like:

.. code-block:: shell

  augur refine \
    --aa --alignment results/HA12/aligned.fasta \
    --metadata results/metadata.tsv --tree results/HA12/tree_raw.nwk \
    --timetree --root best --clock-filter-iqd 1 --date-confidence \
    --branch-length-inference marginal \

>     --output-tree results/HA12/tree.nwk --output-node-data results/HA12/branch_lengths.json

Reviewer: ```suggestion
  augur refine \
    --aa \
    --alignment results/HA12/aligned.fasta \
    --metadata results/metadata.tsv \
    --tree results/HA12/tree_raw.nwk \
    --timetree \
    --root best \
    --clock-filter-iqd 1 \
    --date-confidence \
    --branch-length-inference marginal \
    --output-tree results/HA12/tree.nwk \
    --output-node-data results/HA12/branch_lengths.json
```

Author: Done for most arguments. I find the strict one-per-line a bit too much.

the tree.

``augur ancestral`` can work for AA-only analyses. You will need to provide the following options to reconstruct a single gene/CDS:

* ``--genes <GENE NAME>``
* ``--translations <FILENAME>`` (amino acid FASTA) the filename can be a specific path or include a ``%GENE`` placeholder which will be replaced with the gene name
* [optional] ``--root-sequence <FILENAME>`` (amino acid FASTA) the filename can be a specific path or include a ``%GENE`` placeholder which will be replaced with the gene name
* [optional] ``--annotation <GenBank or GFF>`` (not needed for a single gene analysis)

See :doc:`./root-and-ref-seqs` for more information about how providing a
root-sequence will assign mutations to the root-node of the tree.

The optional annotation file allows the actual nucleotide coordinates of the
gene to be used in the resulting JSON, which slightly changes how the
dataset will appear in Auspice (see :ref:`below <auspice>`).

An example command, where we constructed the tree from a concatenated HA1 + HA2
alignment but where we want to reconstruct mutations on HA1 and HA2 alignments
separately (``results/HA1/aligned.fasta``, ``results/HA2/aligned.fasta``) would
look like:

> .. code-block:: shell
> 
>    augur ancestral \
>      --tree results/HA12/tree.nwk --annotation config/genemap.gff \
>      --translations results/%GENE/aligned.fasta --genes HA1 HA2 \
>      --output-node-data results/HA12/muts.json

Reviewer: ```suggestion
   augur ancestral \
     --tree results/HA12/tree.nwk \
     --annotation config/genemap.gff \
     --translations results/%GENE/aligned.fasta \
     --genes HA1 HA2 \
     --output-node-data results/HA12/muts.json
```

Other commands (filter, export, traits etc)
-------------------------------------------

Other Augur commands in a typical workflow should work with protein sequences, including:
* `augur filter`
* `augur index` (note the index file format is different for AA vs nuc sequences)
* `augur subsample`
* `augur traits`
* `augur export v2`

> Notably, `augur proximity` (and the associated subsampling configs) do not yet work with AA sequences.

Reviewer: Assuming you wanted this to be a bulleted list and the Augur commands to be inline code

```suggestion
Other Augur commands in a typical workflow should work with protein sequences, including:

* ``augur filter``
* ``augur index`` (note the index file format is different for AA vs nuc sequences)
* ``augur subsample``
* ``augur traits``
* ``augur export v2``

Notably, ``augur proximity`` (and the associated subsampling configs) do not yet work with AA sequences.
```

Author: Thanks! RST and me don't get along

* `Auspice version 2.69.0 <https://github.com/nextstrain/auspice/releases/tag/v2.69.0>`__

Changes compared to a nucleotide workflow
=========================================

Indexing, Filtering, and Subsampling
------------------------------------

These commands now handle AA sequences when used with the ``--seq-type aa`` argument.

Note that the index file format is different for AA vs nuc sequences.

Alignment
---------

``augur align --method mafft`` (the default) will work for protein alignments.
Note that Nextclade doesn't support AA inputs and thus can't be used as an aligner in this case.

Tree construction
-----------------

IQ-TREE can be parameterized for AA sequences by adding the the following arguments to
``augur tree``:

* ``--method iqtree`` (the default)
* ``--tree-builder-args '--seqtype AA'``
* ``--substitution-model <MODEL>`` The default IQ-TREE substitution model is ``LG``; see the "Specifying substitution models" section of the `IQ-TREE docs <https://iqtree.github.io/doc/Command-Reference>`__ for a full list of available models.

An example command to construct a tree from a concatenated HA1 + HA2 alignment may look like:

.. code-block:: shell

>    augur tree

Reviewer 2: ```suggestion
   augur tree \
```

     --method iqtree \
     --substitution-model LG \
     --tree-builder-args '--seqtype AA' \
     --alignment results/HA12/aligned.fasta \
     --output results/HA12/tree_raw.nwk

Temporal inference (refine)
---------------------------

Augur refine will handle AA alignments if provided with ``--seq-type aa``.
Internally this changes `TreeTime to use the Jones-Taylor-Thornton (JTT92)
substitution model
<https://treetime.readthedocs.io/en/latest/tutorials/ancestral.html#amino-acid-sequences>`__.
You may also want to use ``--branch-length-inference marginal`` depending on the

> branch lengths involved.

Reviewer 2: Do we know how the use of this argument depends on branch lengths? If we have longer branches with more changes, do we want marginal inference?

Author: Yes, that's my understanding. Joint will pick the best state at each node when traversing the tree whereas marginal will keep the entire profile. For long branches, and especially for AA seqs with ~20 states, the profile is often flat / uncertain, so picking the 'best' isn't clearcut.

Author: Added a sentence to the docs to nod to this

Author: And dove into the code a little more too...

https://github.com/neherlab/treetime/issues/601

Reviewer 2: Ah, that makes sense! Thanks, @jameshadfield.
